File: Retos/Reto #3 - EL GENERADOR DE CONTRASEÑAS [Media]/javascript/almarro1.js
/*
 * Escribe un programa que sea capaz de generar contraseñas de forma aleatoria.
 * Podrás configurar generar contraseñas con los siguientes parámetros:
 * - Longitud: Entre 8 y 16.
 * - Con o sin letras mayúsculas.
 * - Con o sin números.
 * - Con o sin símbolos.
 * (Pudiendo combinar todos estos parámetros entre ellos)
 */

const process = require('process');

// define charsets
const lowercase = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z'];
const uppercase = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'];
const numbers = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '0'];
const symbols = ['\\', '!', '\"', '$', '%', '&', '/', '(', ')', '=', '?', '¿', '@', '#', '¡', '-', '<', '>', '_', '.', ':', ',', ';', '{', '}', '[', ']', '^', '*', 'Ç'];

function generateRandomPassword(length = 8, withNumbers = false, withUppercase = false, withSymbols = false) {
  let charSet = [...lowercase];
  if (withNumbers) {
    charSet = charSet.concat(...numbers)
  }
  if (withUppercase) {
    charSet = charSet.concat(...uppercase)
  }
  if (withSymbols) {
    charSet = charSet.concat(...symbols);
  }

  // iterate on the shuffle to get a better result
  // for (let i = 0; i < 10; i++) {
  for (let i in [...charSet]) {
    charSet.sort(() => 0.5 - Math.random());
  }

  const password = charSet.slice(0, length);
  // verify password fulfills specs
  if (!containsChar(password, lowercase)) {
    return generateRandomPassword(length, withNumbers, withUppercase, withSymbols);
  }
  if (withUppercase && !containsChar(password, uppercase)) {
    return generateRandomPassword(length, withNumbers, withUppercase, withSymbols);
  }
  if (withNumbers && !containsChar(password, numbers)) {
    return generateRandomPassword(length, withNumbers, withUppercase, withSymbols);
  }
  if (withSymbols && !containsChar(password, symbols)) {
    return generateRandomPassword(length, withNumbers, withUppercase, withSymbols);
  }
  return password.join('');
}

/**
 * Function that checks if ```set``` constains any of the elements in ```elements```
 * @param {*} set 
 * @param {*} elements 
 * @returns 
 */
function containsChar(set, elements) {
  return elements.some((c) => set.includes(c));
}

// parse command line args to setup the generator
const arguments = process.argv.slice(2, process.argv.length);
// define defaults
let passLenght = 8;
let useUppercase = false;
let useNumbers = false;
let useSymbols = false;

if (arguments.includes('-h')) {
  console.log(`\
  contraseña -n -l -u -s N
  Si no se porporcionan parámetros, se usarán los parámetros por defecto.
    -n  incluye números
    -u  incluye mayúsculas
    -s  incluye símbolos
    N   número de caracteres a incluir en la constraseña
  `);
  process.exit(0);
}
if (arguments.includes('-u')) { useUppercase = true; }
if (arguments.includes('-n')) { useNumbers = true; }
if (arguments.includes('-s')) { useSymbols = true; }
passLenght = arguments.filter((s) => !s.startsWith('-'))[0] || 8;
if (passLenght < 8 || passLenght > 16) {
  console.log('La longitud debe ser un número entre 8 y 16');
  process.exit(-1);
}

console.log(generateRandomPassword(passLenght, useNumbers, useUppercase, useSymbols));

File: Retos/Reto #3 - EL GENERADOR DE CONTRASEÑAS [Media]/python/almarro1.py
"""
 * Escribe un programa que sea capaz de generar contraseñas de forma aleatoria.
 * Podrás configurar generar contraseñas con los siguientes parámetros:
 * - Longitud: Entre 8 y 16.
 * - Con o sin letras mayúsculas.
 * - Con o sin números.
 * - Con o sin símbolos.
 * (Pudiendo combinar todos estos parámetros entre ellos)
"""

import random
import sys

arguments = sys.argv[1:]

lowercaseAlphabet = 'abcdefghijklmnopqrstuvwxyz'
uppercaseAlphabet = lowercaseAlphabet.upper()
numbers = '1234567890'
symbols = '\\!|"@·#$%&/()=?¿\'¡<>,;.:-_ç`+^*¨}{[]'


def password_generator(length: int = 8, with_numbers: bool = False, with_upper: bool = False, with_symbols: bool = False) -> str:
    alphabet = lowercaseAlphabet
    if with_numbers:
        alphabet += numbers
    if with_upper:
        alphabet += uppercaseAlphabet
    if with_symbols:
        alphabet += symbols

    characters = [letter for letter in alphabet]
    random.shuffle(characters)
    password = "".join(characters[0:length])
    # TODO: falta verificar que se cumplan las especificaciones
    while not verify_specs(password, with_numbers, with_upper, with_symbols):
        random.shuffle(characters)
        password = "".join(characters[0:length])
    return password


def verify_specs(password: str, with_numbers: bool, with_upper: bool, with_symbols: bool):
    _lower = [letter for letter in lowercaseAlphabet]
    for letter in password:
        if letter in _lower:
            break
    else:
        # print('Sin minúsculas')
        return False
    if with_numbers:
        _numbers = [number for number in numbers]
        for letter in password:
            if letter in _numbers:
                break
        else:
            # print('Sin números')
            return False
    if with_upper:
        _upper = [letter for letter in uppercaseAlphabet]
        for letter in password:
            if letter in _upper:
                break
        else:
            # print('Sin mayúsculas')
            return False
    if with_symbols:
        _symbols = [symbol for symbol in symbols]
        for letter in password:
            if letter in _symbols:
                break
        else:
            # print('Sin símbolos')
            return False

    return True


with_upper = False
with_symbols = False
with_numbers = False
length = 8

if '-h' in arguments:
    print('''\
  comando -h -s -n -u N
      -h  Muestra esta ayuda
      -s  Inserta símbolos
      -n  Inserta números
      -u  Inserta mayúsculas
       N  Número de caracteres. Debe ser un número entre 8 y 16
    ''')
    sys.exit(0)

if '-u' in arguments:
    with_upper = True
    arguments.remove('-u')

if '-s' in arguments:
    with_symbols = True
    arguments.remove('-s')

if '-n' in arguments:
    with_numbers = True
    arguments.remove('-n')

if len(arguments) > 1:
    print('Demasiados parámetros')
    sys.exit(-1)
try:
    length = int(arguments[0]) if len(arguments) == 1 else 8
    if length < 8 or length > 16:
        print(
            f'La longitud debe ser un número entre 8 y 16. Especificaste {length}')
        sys.exit(-1)
except Exception:
    print(f'Longitud errónea {arguments[0]}')
    sys.exit(-1)

print(
    f'Ejecutando con longitud={length}, números={with_numbers}, mayúsculas={with_upper}, símbolos={with_symbols}')
print(password_generator(length, with_numbers, with_upper, with_symbols))
